clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -324,18 +339,91 @@ public ObjectLocation pathToObjectLocation(Path path) {
         }
 
         URI uri = path.toUri();
-
         ObjectLocation loc = new ObjectLocation();
         loc.setRepository(uri.getHost());
         // extract ref and rest of the path after removing the '/' prefix
         String s = ObjectLocation.trimLeadingSlash(uri.getPath());
-        int i = s.indexOf(Constants.SEPARATOR);
+        int i = s.indexOf(Constants.URI_SEPARATOR);
         if (i == -1) {
             loc.setRef(s);
+            loc.setPath("");
         } else {
             loc.setRef(s.substring(0, i));
             loc.setPath(s.substring(i+1));
         }
         return loc;
     }
+
+    class ListingIterator implements RemoteIterator<LocatedFileStatus> {
+        private final URI uri;
+        private final ObjectLocation objectLocation;
+        private final String delimiter;
+        private final int amount;
+        private String nextOffset;
+        private boolean last;
+        private List<ObjectStats> chunk;
+        private int pos;
+
+        public ListingIterator(Path path, boolean recursive, int amount) {
+            this.uri = path.toUri();
+            this.chunk = Collections.emptyList();
+            this.objectLocation = pathToObjectLocation(path);
+            String locationPath = this.objectLocation.getPath();
+            if (locationPath.length() > 0 && !locationPath.endsWith(URI_SEPARATOR)) {
+                this.objectLocation.setPath(locationPath + URI_SEPARATOR);
+            }
```

Comment:
1. If this is the desired behaviour, shouldn't it be in `pathToObjectLocation`?
2. I'm not sure that this is correct.  Does Hadoop allow listing a single file? (Can open an issue to check and maybe fix this, rather than block this PR).